src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -459,13 +382,66 @@ private void DetachFromDocument(uint docCookie, string moniker)
             }
         }
 
+        private void SubscribeToRunningDocTableEvents()
+        {
+            _runningDocumentTableEventSink.OnRenameDocument += HandleRenameDocumentEvent;
+            _runningDocumentTableEventSink.OnReloadDocumentData += HandleReloadDocumentEvent;
+            _runningDocumentTableEventSink.OnInitializedDocument += HandleInitializedDocumentEvent;
+            _runningDocumentTableEventSink.OnCloseDocument += HandleCloseDocumentEvent;
+        }
+
+        private void UnSubscribeFromRunningDocTableEvents()
+        {
+            _runningDocumentTableEventSink.OnRenameDocument -= HandleRenameDocumentEvent;
+            _runningDocumentTableEventSink.OnReloadDocumentData -= HandleReloadDocumentEvent;
+            _runningDocumentTableEventSink.OnInitializedDocument -= HandleInitializedDocumentEvent;
+            _runningDocumentTableEventSink.OnCloseDocument -= HandleCloseDocumentEvent;
+        }
+
+        private void HandleRenameDocumentEvent(object sender, RunningDocumentTableRenamedEventArgs args)
+        {
+            // We want to consider this file to be added in one of two situations:
+            //
+            // 1) the old file already was a misc file, at which point we might just be doing a rename from
+            //    one name to another with the same extension
+            // 2) the old file was a different extension that we weren't tracking, which may have now changed
+            if (TryUntrackClosingDocument(args.DocCookie, args.OldMoniker) || TryGetLanguageInformation(args.OldMoniker) == null)
+            {
+                if (_runningDocumentTableEventSink.TryGetBuffer(args.DocCookie, out var buffer))
+                {
+                    // Add the new one, if appropriate.
+                    TrackOpenedDocument(args.DocCookie, args.Moniker, buffer);
+                }
+            }
+        }
+
+        private void HandleReloadDocumentEvent(object sender, RunningDocumentTableEventArgs args)
+        {
+            if (args.Moniker != null && TryGetLanguageInformation(args.Moniker) != null && !_docCookiesToProjectIdAndContainer.ContainsKey(args.DocCookie))
```

Comment:
why only this has container check?

Reply:
when moniker could be null? TryGetLanguageInformation check could be skipped?

so, docCookiesToProjectIdAndContainer will contain doc cookie if the file belong to misc project. so, basically what it is saying is if the file belong to another workspace, but related to one of roslyn document, track open document... but since TraceOpenedDocument already has this
if (!_docCookieToWorkspaceRegistration.ContainsKey(docCookie))
what it really mean is that, this document is in other workspace without ever assigned to misc workspace first, then we track it now.

now question is, is that possible? can reload happen without ever go through misc workspace first?

I guess it is fine as defensive mechanism in case some assumption get broken?

Reply:
> 
> 
> when moniker could be null? TryGetLanguageInformation check could be skipped?
> 

Good question.  I haven't been able to get a case so far, but @jasonmalinowski might know why this check was here?  It was copied from the existing implementation.

>
>
> now question is, is that possible? can reload happen without ever go through misc workspace first?

Haven't encountered this either in my testing.  @jasonmalinowski do we need this check? 

In absence of other evidence, I would prefer to leave these checks there since that is what the original misc workspace is doing.